### PR TITLE
Feature/filter geom

### PR DESCRIPTION
**Pull request by andrewstanfordjason**
This PR adds filter2DGeometry classes and associated tests and support classes. It's rather large and could do with being pulled apart but also comes with 1.7K tests.

### CODE REVIEW

File: lib_nn/api/geom/Filter2dGeometry.hpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+   *
+   * Two filter geometries are equal iff each of their component geometries
+   * (input, output, window) are equal.
+   */
+  bool operator==(Filter2dGeometry other) const;
+
+  /**
+   * Test for inequality of two filter geometries.
+   *
+   * Two filter geometries are unequal iff any of their component geometries
```
Comment (xross) [inline lib_nn/api/geom/Filter2dGeometry.hpp:83]: Comment typo "iff" (or unexpected if and only if.. which is a bit a**l)

File: lib_nn/api/geom/Filter2dGeometry.hpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+   * the receptive field itself_ which are beyond the input image's bounds (i.e.
+   * the "gaps" due to dilation do not count). This function indicates the
+   * padding required if the input image is to _actually be padded in memmory_.
+   */
+  padding_t Padding() const;
+
+  /// TODO: astew: Are these correct? Does the receptive volume of a depthwise
+  /// filter get
+  ///       multiplied by the number of input channels? I suspect that the
+  ///       following functions are unnecessary, as for a properly-speified
```
Comment (xross) [inline lib_nn/api/geom/Filter2dGeometry.hpp]: Remove?

File: lib_nn/src/cpp/filt2d/geom/Filter2dGeometry.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  padding.MakeUnsigned();
+  return padding;
+}
+
+bool Filter2dGeometry::IsDepthwise() const {
+  // A model is depthwise if the window channel stride is 1 and its depth is 1.
+  // If stride is 0, then it is 'dense'. If stride is not 0 or 1, or if the
+  // stride is 1 and the depth is not 0, the behavior is undefined.
+  return window.stride.channel == 1 && window.shape.depth == 1;
+}
```
Comment (xross) [inline lib_nn/src/cpp/filt2d/geom/Filter2dGeometry.cpp]: GitHub complaint about no newline at end of file

File: lib_nn/src/cpp/filt2d/geom/WindowLocation.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+int WindowLocation::InputIndex(const int filter_row, const int filter_col,
+                               const int filter_chan) const {
+  assert(filter_row >= 0);
+  assert(filter_row < filter.window.shape.height);
+  assert(filter_col >= 0);
+  assert(filter_col < filter.window.shape.width);
+  assert(filter_chan >= 0);
+  assert(filter_chan < filter.window.shape.depth);
```
Comment (xross) [inline lib_nn/src/cpp/filt2d/geom/WindowLocation.cpp:42]: Is there a reason we mix the terms "chan" and "depth"? It would be nice if it was consistent,

**Review by xross — APPROVED** [review]
Looks good. I only skimmed the test code.

File: lib_nn/api/geom/Filter2dGeometry.hpp (hunk 1 above)
Comment (andrewstanfordjason) [reply lib_nn/api/geom/Filter2dGeometry.hpp:83]: I'm ok with this

File: lib_nn/src/cpp/filt2d/geom/WindowLocation.cpp (hunk 4 above)
Comment (andrewstanfordjason) [reply lib_nn/src/cpp/filt2d/geom/WindowLocation.cpp:42]: this is a good point

Comment (andrewstanfordjason) [reply lib_nn/src/cpp/filt2d/geom/WindowLocation.cpp:42]: im going to leave this for now

**Pull request merged**